Libraries/DevBundleLoading/loadBundle.js

```diff
@@ -0,0 +1,149 @@
+    dataListener = Networking.addListener(
+      'didReceiveNetworkData',
+      ([requestId, response]) => {
+        if (requestId === id) {
+          responseText = response;
+        }
+      },
+    );
+    responseListener = Networking.addListener(
+      'didReceiveNetworkResponse',
+      ([requestId, status, responseHeaders]) => {
+        if (requestId === id) {
+          headers = responseHeaders;
+        }
+      },
+    );
+    completeListener = Networking.addListener(
+      'didCompleteNetworkResponse',
+      ([requestId, error]) => {
+        if (requestId === id) {
+          if (error) {
+            reject(error);
+          } else {
+            resolve({body: responseText, headers});
+          }
+        }
+      },
+    );
+    Networking.sendRequest(
+      'GET',
+      'asyncRequest',
+      url,
+      {},
+      '',
+      'text',
+      false,
+      0,
+      requestId => {
+        id = requestId;
+      },
+      true,
+    );
+    //$FlowFixMe[incompatible-return]
+  }).finally(() => {
+    dataListener && dataListener.remove();
+    completeListener && completeListener.remove();
+    responseListener && responseListener.remove();
+  });
+}
+
+function buildUrlForBundle(bundlePath, params) {
+  const {
+    fullBundleUrl,
+    url: serverUrl,
+    bundleLoadedFromServer,
+  } = getDevServer();
+  if (!bundleLoadedFromServer) {
+    throw new Error(
+      'This bundle was compiled with transformer.experimentalImportBundleSupport and can only be used when connected to a Metro server.',
+    );
+  }
+  let query = {};
+  if (fullBundleUrl != null) {
+    const queryStart = fullBundleUrl.indexOf('?');
+    if (queryStart !== -1) {
+      query = qs.parse(fullBundleUrl.substring(queryStart + 1));
+    }
+  }
+  Object.assign(query, params);
+  return serverUrl + bundlePath + '.bundle?' + qs.stringify(query);
+}
+
+module.exports = function(bundlePath: string): Promise<mixed> {
+  if (NativeDevSplitBundleLoader && NativeDevSplitBundleLoader.loadBundle) {
+    return NativeDevSplitBundleLoader.loadBundle(bundlePath).catch(e => {
+      // On Android 'e' is not an instance of Error, which seems to be a bug.
+      // As a workaround, re-throw an Error to not break the error handling code.
+      throw new Error(e.message);
+    });
+  }
+
+  const requestUrl = buildUrlForBundle(bundlePath, {
+    modulesOnly: 'true',
+    runModule: 'false',
+    // The JavaScript loader does not support bytecode.
+    runtimeBytecodeVersion: null,
+  });
+
+  LoadingView.showMessage('Downloading...', 'load');
+  return asyncRequest(requestUrl)
+    .then(({body, headers}) => {
+      if (
+        headers['Content-Type'] != null &&
+        headers['Content-Type'].indexOf('application/json') >= 0
+      ) {
+        // Errors are returned as JSON.
+        throw new Error(
+          JSON.parse(body).message || `Unknown error fetching '${bundlePath}'`,
+        );
+      }
+
+      HMRClient.registerBundle(requestUrl);
```

Comment:
fyi I found a bug with this code that you also have at Facebook. When Metro's websocket disconnects and reconnects, it will not replay the `registerBundle` calls and Metro won't update the right files any longer. Most likely it will manifest in Fast Refresh breaking until you reload. The fix is to append `registerBundle` calls to a `Set` and replay them when reconnecting.